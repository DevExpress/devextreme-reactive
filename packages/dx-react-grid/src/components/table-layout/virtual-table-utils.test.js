import {
  getVisibleBoundary,
  getSpanBoundary,
  collapseBoundaries,
  getCollapsedColumns,
  getCollapsedCells,
  getCollapsedGrid,
  STUB_TYPE,
} from './virtual-table-utils';

describe('VirtualTableLayout utils', () => {
  describe('#getVisibleBoundary', () => {
    it('should work in the simplest case', () => {
      const items = [
        { size: 40 },
        { size: 40 },
        { size: 40 },
        { size: 40 },
        { size: 40 },
        { size: 40 },
        { size: 40 },
      ];

      expect(getVisibleBoundary(items, 80, 120, item => item.size))
        .toEqual([2, 4]);
    });

    it('should work with overscan', () => {
      const items = [
        { size: 40 },
        { size: 40 },
        { size: 40 },
        { size: 40 },
        { size: 40 },
        { size: 40 },
        { size: 40 },
      ];

      expect(getVisibleBoundary(items, 80, 120, item => item.size, 1))
        .toEqual([1, 5]);
    });
  });

  describe('#getSpanBoundary', () => {
    it('should work with span before visible area', () => {
      const items = [
        { colSpan: 1 },
        { colSpan: 2 },
        { colSpan: 1 },
        { colSpan: 1 },
        { colSpan: 1 },
        { colSpan: 1 },
        { colSpan: 1 },
      ];

      expect(getSpanBoundary(items, [2, 4], item => item.colSpan))
        .toEqual([1, 4]);
    });

    it('should work with span after visible area', () => {
      const items = [
        { colSpan: 1 },
        { colSpan: 1 },
        { colSpan: 1 },
        { colSpan: 1 },
        { colSpan: 2 },
        { colSpan: 1 },
        { colSpan: 1 },
      ];

      expect(getSpanBoundary(items, [2, 4], item => item.colSpan))
        .toEqual([2, 5]);
    });

    it('should work with span greater than visible area', () => {
      const items = [
        { colSpan: 1 },
        { colSpan: 5 },
        { colSpan: 1 },
        { colSpan: 1 },
        { colSpan: 1 },
        { colSpan: 1 },
        { colSpan: 1 },
      ];

      expect(getSpanBoundary(items, [2, 4], item => item.colSpan))
        .toEqual([1, 5]);
    });
  });

  describe('#collapseBoundaries', () => {
    it('should work with spans before visible area', () => {
      const itemsCount = 9;
      const visibleBoundary = [3, 5];
      const spanBoundaries = [
        [1, 5],
        [2, 5],
      ];

      expect(collapseBoundaries(itemsCount, visibleBoundary, spanBoundaries))
        .toEqual([
          [0, 0], // stub
          [1, 1], // stub
          [2, 2], // stub
          [3, 3], // visible
          [4, 4], // visible
          [5, 5], // visible
          [6, 8], // stub
        ]);
    });

    it('should work with spans after visible area', () => {
      const itemsCount = 9;
      const visibleBoundary = [3, 5];
      const spanBoundaries = [
        [3, 7],
        [3, 6],
      ];

      expect(collapseBoundaries(itemsCount, visibleBoundary, spanBoundaries))
        .toEqual([
          [0, 2], // stub
          [3, 3], // visible
          [4, 4], // visible
          [5, 5], // visible
          [6, 6], // stub
          [7, 7], // stub
          [8, 8], // stub
        ]);
    });

    it('should work with spans greater than visible area', () => {
      const itemsCount = 9;
      const visibleBoundary = [3, 5];
      const spanBoundaries = [
        [1, 6],
        [2, 7],
      ];

      expect(collapseBoundaries(itemsCount, visibleBoundary, spanBoundaries))
        .toEqual([
          [0, 0], // stub
          [1, 1], // stub
          [2, 2], // stub
          [3, 3], // visible
          [4, 4], // visible
          [5, 5], // visible
          [6, 6], // stub
          [7, 7], // stub
          [8, 8], // stub
        ]);
    });
  });

  describe('#getCollapsedColumns', () => {
    it('should work', () => {
      const columns = [
        { key: 0, width: 40 },
        { key: 1, width: 40 },
        { key: 2, width: 40 },
        { key: 3, width: 40 },
        { key: 4, width: 40 },
        { key: 5, width: 40 },
        { key: 6, width: 40 },
        { key: 7, width: 40 },
      ];
      const visibleBoundary = [3, 4];
      const boundaries = [[0, 0], [1, 2], [3, 3], [4, 4], [5, 6], [7, 7]];
      const getColumnWidth = column => column.width;

      const result = [
        { type: 'stub', key: 'stub_0_0', width: 40 },
        { type: 'stub', key: 'stub_1_2', width: 80 },
        { ...columns[3] },
        { ...columns[4] },
        { type: 'stub', key: 'stub_5_6', width: 80 },
        { type: 'stub', key: 'stub_7_7', width: 40 },
      ];

      expect(getCollapsedColumns(columns, visibleBoundary, boundaries, getColumnWidth))
        .toEqual(result);
    });
  });

  describe('#getCollapsedCells', () => {
    it('should work', () => {
      const columns = [
        { key: 0, colSpan: 1 },
        { key: 1, colSpan: 6 },
        { key: 2, colSpan: 1 },
        { key: 3, colSpan: 1 },
        { key: 4, colSpan: 1 },
        { key: 5, colSpan: 1 },
        { key: 6, colSpan: 1 },
        { key: 7, colSpan: 1 },
      ];
      const spanBoundary = [1, 6];
      const boundaries = [[0, 0], [1, 2], [3, 3], [4, 4], [5, 6], [7, 7]];
      const getColSpan = column => column.colSpan;

      const result = [
        { column: { type: 'stub', key: 'stub_0_0' }, colSpan: 1 },
        { column: columns[1], colSpan: 4 },
        { column: columns[3], colSpan: 1 },
        { column: columns[4], colSpan: 1 },
        { column: columns[5], colSpan: 1 },
        { column: { type: 'stub', key: 'stub_7_7' }, colSpan: 1 },
      ];

      expect(getCollapsedCells(columns, spanBoundary, boundaries, getColSpan))
        .toEqual(result);
    });
  });

  describe('#getCollapsedGrid', () => {
    it('should work in simple case', () => {
      const args = {
        rows: [
          { key: 0, height: 40 },
          { key: 1, height: 40 }, // visible (overscan)
          { key: 2, height: 40 }, // visible (overscan)
          { key: 3, height: 40 }, // visible (overscan)
          { key: 4, height: 40 }, // visible
          { key: 5, height: 40 }, // visible (overscan)
          { key: 6, height: 40 }, // visible (overscan)
          { key: 7, height: 40 }, // visible (overscan)
          { key: 8, height: 40 },
        ],
        columns: [
          { key: 0, width: 40 },
          { key: 1, width: 40 }, // visible (overscan)
          { key: 2, width: 40 }, // visible
          { key: 3, width: 40 }, // visible (overscan)
          { key: 4, width: 40 },
        ],
        top: 160,
        left: 80,
        height: 40,
        width: 40,
      };

      const result = getCollapsedGrid(args);

      expect(result.rows.map(row => row.row.type))
        .toEqual([STUB_TYPE, ...Array.from({ length: 7 }).map(() => undefined), STUB_TYPE]);
      expect(result.columns.map(column => column.type))
        .toEqual([STUB_TYPE, ...Array.from({ length: 3 }).map(() => undefined), STUB_TYPE]);
      expect(result.rows[1].cells.map(cell => cell.colSpan))
        .toEqual([...Array.from({ length: 5 }).map(() => 1)]);
    });

    it('should return empty resule when there are no columns', () => {
      const args = {
        rows: [],
        columns: [
          { key: 0, width: 40 },
        ],
        top: 0,
        left: 0,
        height: 80,
        width: 80,
      };
      const result = {
        columns: [],
        rows: [],
      };
      expect(getCollapsedGrid(args))
        .toEqual(result);
    });

    it('should return empty resule when there are no rows', () => {
      const args = {
        rows: [
          { key: 0, height: 40 },
        ],
        columns: [],
        top: 0,
        left: 0,
        height: 80,
        width: 80,
      };
      const result = {
        columns: [],
        rows: [],
      };
      expect(getCollapsedGrid(args))
        .toEqual(result);
    });

    it('should work with collspan', () => {
      const args = {
        rows: [
          { key: 0, height: 40 }, // visible
          { key: 1, height: 40 }, // visible (overscan)
          { key: 2, height: 40 }, // visible (overscan)
          { key: 3, height: 40 }, // visible (overscan)
          { key: 4, height: 40 },
        ],
        columns: [
          { key: 0, width: 40 }, // stub ┐
          { key: 1, width: 40 }, // stub ┘
          { key: 2, width: 40 }, // stub
          { key: 3, width: 40 }, // visible (overscan)
          { key: 4, width: 40 }, // visible
          { key: 5, width: 40 }, // visible (overscan)
          { key: 6, width: 40 }, // stub ┐
          { key: 7, width: 40 }, // stub ┘
          { key: 8, width: 40 }, // stub
        ],
        top: 0,
        left: 160,
        height: 40,
        width: 40,
        getColSpan: (row, column) => {
          if (row.key === 0 && column.key === 2) return 2;
          if (row.key === 0 && column.key === 5) return 3;
          if (row.key === 1 && column.key === 2) return 6;
          if (row.key === 2 && column.key === 0) return 9;
          return 1;
        },
      };

      const result = getCollapsedGrid(args);

      expect(result.columns.map(column => column.type))
        .toEqual([
          STUB_TYPE, STUB_TYPE,
          ...Array.from({ length: 3 }).map(() => undefined),
          STUB_TYPE, STUB_TYPE,
        ]);

      expect(result.rows[0].cells.map(cell => cell.colSpan))
        .toEqual([1, 2, 1, 1, 2, 1, 1]);
      expect(result.rows[1].cells.map(cell => cell.colSpan))
        .toEqual([1, 5, 1, 1, 1, 1, 1]);
      expect(result.rows[2].cells.map(cell => cell.colSpan))
        .toEqual([7, 1, 1, 1, 1, 1, 1]);
    });
  });
});
