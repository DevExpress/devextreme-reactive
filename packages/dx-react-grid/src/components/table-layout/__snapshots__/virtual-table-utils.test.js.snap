// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VirtualTableLayout utils #getCollapsedGrid should work in simple case 1`] = `
Object {
  "columns": Array [
    Object {
      "key": "stub_0_0",
      "type": "stub",
      "width": 40,
    },
    Object {
      "column": Object {
        "key": 1,
        "width": 40,
      },
      "type": "visible",
      "width": 40,
    },
    Object {
      "column": Object {
        "key": 2,
        "width": 40,
      },
      "type": "visible",
      "width": 40,
    },
    Object {
      "column": Object {
        "key": 3,
        "width": 40,
      },
      "type": "visible",
      "width": 40,
    },
    Object {
      "key": "stub_4_5",
      "type": "stub",
      "width": 80,
    },
  ],
  "rows": Array [
    Object {
      "height": 40,
      "key": "stub_0_0",
      "type": "stub",
    },
    Object {
      "cells": Array [
        Object {
          "colSpan": 1,
          "key": "stub_0_0",
          "type": "stub",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 1,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 2,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 3,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "key": "stub_4_5",
          "type": "stub",
        },
      ],
      "height": 40,
      "row": Object {
        "height": 40,
        "key": 1,
      },
      "type": "visible",
    },
    Object {
      "cells": Array [
        Object {
          "colSpan": 1,
          "key": "stub_0_0",
          "type": "stub",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 1,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 2,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 3,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "key": "stub_4_5",
          "type": "stub",
        },
      ],
      "height": 40,
      "row": Object {
        "height": 40,
        "key": 2,
      },
      "type": "visible",
    },
    Object {
      "cells": Array [
        Object {
          "colSpan": 1,
          "key": "stub_0_0",
          "type": "stub",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 1,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 2,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 3,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "key": "stub_4_5",
          "type": "stub",
        },
      ],
      "height": 40,
      "row": Object {
        "height": 40,
        "key": 3,
      },
      "type": "visible",
    },
    Object {
      "cells": Array [
        Object {
          "colSpan": 1,
          "key": "stub_0_0",
          "type": "stub",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 1,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 2,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 3,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "key": "stub_4_5",
          "type": "stub",
        },
      ],
      "height": 40,
      "row": Object {
        "height": 40,
        "key": 4,
      },
      "type": "visible",
    },
    Object {
      "cells": Array [
        Object {
          "colSpan": 1,
          "key": "stub_0_0",
          "type": "stub",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 1,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 2,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 3,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "key": "stub_4_5",
          "type": "stub",
        },
      ],
      "height": 40,
      "row": Object {
        "height": 40,
        "key": 5,
      },
      "type": "visible",
    },
    Object {
      "cells": Array [
        Object {
          "colSpan": 1,
          "key": "stub_0_0",
          "type": "stub",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 1,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 2,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 3,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "key": "stub_4_5",
          "type": "stub",
        },
      ],
      "height": 40,
      "row": Object {
        "height": 40,
        "key": 6,
      },
      "type": "visible",
    },
    Object {
      "cells": Array [
        Object {
          "colSpan": 1,
          "key": "stub_0_0",
          "type": "stub",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 1,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 2,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 3,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "key": "stub_4_5",
          "type": "stub",
        },
      ],
      "height": 40,
      "row": Object {
        "height": 40,
        "key": 7,
      },
      "type": "visible",
    },
    Object {
      "height": 40,
      "key": "stub_8_8",
      "type": "stub",
    },
  ],
}
`;

exports[`VirtualTableLayout utils #getCollapsedGrid should work with collspan 1`] = `
Object {
  "columns": Array [
    Object {
      "key": "stub_0_0",
      "type": "stub",
      "width": 40,
    },
    Object {
      "key": "stub_1_1",
      "type": "stub",
      "width": 40,
    },
    Object {
      "column": Object {
        "key": 2,
        "width": 40,
      },
      "type": "visible",
      "width": 40,
    },
    Object {
      "column": Object {
        "key": 3,
        "width": 40,
      },
      "type": "visible",
      "width": 40,
    },
    Object {
      "column": Object {
        "key": 4,
        "width": 40,
      },
      "type": "visible",
      "width": 40,
    },
    Object {
      "key": "stub_5_7",
      "type": "stub",
      "width": 120,
    },
  ],
  "rows": Array [
    Object {
      "cells": Array [
        Object {
          "colSpan": 1,
          "key": "stub_0_0",
          "type": "stub",
        },
        Object {
          "colSpan": 1,
          "key": "stub_1_1",
          "type": "stub",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 2,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 2,
          "column": Object {
            "key": 3,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 4,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "key": "stub_5_7",
          "type": "stub",
        },
      ],
      "height": 40,
      "row": Object {
        "height": 40,
        "key": 0,
      },
      "type": "visible",
    },
    Object {
      "cells": Array [
        Object {
          "colSpan": 1,
          "key": "stub_0_0",
          "type": "stub",
        },
        Object {
          "colSpan": 3,
          "column": Object {
            "key": 1,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 2,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 3,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 4,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "key": "stub_5_7",
          "type": "stub",
        },
      ],
      "height": 40,
      "row": Object {
        "height": 40,
        "key": 1,
      },
      "type": "visible",
    },
    Object {
      "cells": Array [
        Object {
          "colSpan": 4,
          "column": Object {
            "key": 0,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 1,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 2,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 3,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 4,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "key": "stub_5_7",
          "type": "stub",
        },
      ],
      "height": 40,
      "row": Object {
        "height": 40,
        "key": 2,
      },
      "type": "visible",
    },
    Object {
      "cells": Array [
        Object {
          "colSpan": 1,
          "key": "stub_0_0",
          "type": "stub",
        },
        Object {
          "colSpan": 1,
          "key": "stub_1_1",
          "type": "stub",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 2,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "column": Object {
            "key": 3,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 2,
          "column": Object {
            "key": 4,
            "width": 40,
          },
          "type": "visible",
        },
        Object {
          "colSpan": 1,
          "key": "stub_5_7",
          "type": "stub",
        },
      ],
      "height": 40,
      "row": Object {
        "height": 40,
        "key": 3,
      },
      "type": "visible",
    },
    Object {
      "height": 40,
      "key": "stub_4_4",
      "type": "stub",
    },
  ],
}
`;
